test/build_container.sh

```diff
@@ -26,6 +26,7 @@ elif [ "$RESTY_IMAGE_BASE" == "alpine" ]; then
   cp output/*.${ARCHITECTURE}.apk.tar.gz docker-kong/alpine/kong.tar.gz
   BUILD_DIR="alpine"
 elif [ "$RESTY_IMAGE_BASE" == "centos" ] || [ "$RESTY_IMAGE_BASE" == "amazonlinux" ]; then
```

Comment:
It seems we still use `$RESTY_IMAGE_BASE` instead of `$PACKAGE_TYPE` is still being used in the tests, isn't that problematic?

Reply:
It seems problematic: I was going to ask why removing the `sed` line was needed in (we remove them for for testing amazonlinux assets, says the commit message).

Reply:
when testing an asset we've produced we use git://kong/docker-kong as a test harness, however the Dockerfile's there in some cases are close to a correct harness less some differences. Because the docker-kong Dockerfile's are used for our official docker image builds there's some restrictions that don't allow them to be as flexible as we require.

All of that is a long way of saying I HATE the spots we employ `sed` but haven't found a suitable alternative.

The other spot we will still require RESTY_IMAGE_BASE and RESTY_IMAGE_TAG is in the fpm script as it needs to capture differences that occur between OS versions such as a dependent package was renamed etc

Reply:
`sed` is convenient _in the moment_: it gets the job done, but it often requires you to have the whole "tree of dependencies" inside your head. `sed` gives trouble _in the future_, when someone who doesn't have the "tree of dependencies" fresh in their heads (can be future you) tries to make sense of it.

> haven't found a suitable alternative.

It might be that you are looking for an alternative that is convenient both _in the moment_ and _in the future_.

I think we should consider using a solution that makes things more maintainable - even if it makes things more complex in the short term. Usually this means putting the "dependencies" out of our heads and making them explicit in the code.

A recent example I can think of: In docker-kong, we use `sed` inside our update.sh script to get the download urls of each distro, directly from their Dockerfiles. The problem with that is that that's implicit knowledge. If you look at the Ubuntu Dockerfile, you cannot know that `update.sh` depends on it - that the "source of truth for the ubuntu download url" is the Dockerfile itself. So you might [change the url calculation in a way that breaks the sed in update.sh, and not notice anything](https://github.com/Kong/docker-kong/pull/537/commits/34e715e8eb060650f6f28954272c454b7aa11981). This implicit knowledge makes things difficult to maintain.

The alternative,is making those dependencies more explicit. We could introduce a file of "shared globals". Which could contain stuff like:

```
# globals.sh

def ubuntu_download_url() {
  KONG_VERSION=$1
  UBUNTU_CODENAME=$2
  echo "https://download.konghq.com/gateway-${KONG_VERSION%%.*}.x-
ubuntu-${UBUNTU_CODENAME}/pool/all/k/kong/kong_${KONG_VERSION}_amd64.deb"
}

def centos_download_url(){
  ...
}
```
(This is just an example, the exact way in which we share stuff might be different).

Once we have that we could use`globals.sh` in both `update.sh` and when setting parameters for the `Dockerfile`. This would make dependencies more explicit: the inclusion in `update.sh` would be direct. The Dockerfile would need a `DOWNLOAD_URL` ENV, which would be provided by a `make_dockerfile.sh` script, that would include `globals.sh`.

But that is way more complicated than using a simple `sed` inside update.sh. But that is the tradeoff: In order to avoid problems in the future we have to make sacrifices in the present.